[Electrophysiology Browser] New module

cmadjar opened the pull request:
## Brief summary of changes

This PR includes SQL changes and code required for the new Electrophysiology Browser module. 

Screenshot of the menu filter page (reactified and data frameworked):
![Screen Shot 2019-09-17 at 5 20 43 PM](https://user-images.githubusercontent.com/1402456/65080730-f403d580-d96f-11e9-9403-ec49e7a65897.png)

Screenshot of the view session page:
![Screen Shot 2019-09-17 at 5 21 07 PM](https://user-images.githubusercontent.com/1402456/65080748-ffef9780-d96f-11e9-98bf-8e88b8b5fb2c.png)

### IMPORTANT

This has been a wonderful team work and I want the authors to be recognized as most of the work to create the module has been done by them :):
- @um4r12 for the original back end PHP code
- @maltheism for the react front end code
- @christinerogers, Derek for the UI and design
- @cmadjar, upgrade of the code to 21.0, to use the latest react for the menu filter, to use the data framework for the menu filter, few minor bug fixes, design, SQL etc...


#### Testing instructions (if applicable)

1. First, make sure to reload the latest RaisinBread that include EEG datasets (should be included in this branch since I rebased it after the RB dataset got merged)
2. Go to the Electrophysiology menu tab and select the Electrophysiology Browser to direct you to the new module. 
3. Have fun clicking and checking the pages :) 

#### Links to related tickets (GitHub, Redmine, ...)

* https://github.com/aces/Loris/issues/5031

cmadjar commented on htdocs/mri/jiv/get_file.php:
```diff
@@ -123,6 +123,26 @@
     $MimeType         = 'application/xml';
     $DownloadFilename = basename($File);
     break;
+case 'edf':
```
@johnsaigle let me know if those modifications of the get_file.php file is not correct.

johnsaigle replied:
I think this looks fine. 👍

christinerogers replied:
based on my experience with EDF, there may be an additional modification to the bottom of this file to facilitate large downloads. 
- [ ] Note to self @christinerogers to come back to this

cmadjar commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,35 @@
+# Electrophysiology Browser
```
@christinerogers let me know if you have any feedback on the README.md (or other files too ;))

christinerogers replied:
This looks like a good first draft. I'll think later about what it might be missing. 
The only change I can suggest (and I'll make this commit myself) -- change "Viewing" to "Visualization"

christinerogers replied:
We should also add: (this is a note to self) 
- [ ] Download -- all the different great things you can download! (and who has permission to do it?  Site-based?

cmadjar replied:
You are so right!! I totally forgot... I will add it in the next commit.

johnsaigle commented on .gitignore:
```diff
@@ -1,11 +1,13 @@
 project
 smarty/templates_c
+.gitignore
```
I don't think these should be ignored. This file and the `.github/` folder are both tracked in the repo.

cmadjar replied:
ooops. will be corrected in the next commit.

johnsaigle commented on SQL/0000-00-02-Menus.sql:
```diff
@@ -35,9 +35,10 @@ INSERT INTO LorisMenu (Label, OrderNumber) VALUES
-     ('Reports', 4),
-     ('Tools', 5),
-     ('Admin', 6);
+     ('Electrophysiology', 4)
```
Should this be under Imaging?

cmadjar replied:
Actually, this is a new modality altogether not really considered as imaging but as electrophysiology (at least this is my understanding, @christinerogers what do you think?)

johnsaigle replied:
Cool, just checking. I definitely don't know.

christinerogers replied:
You're both right, and I remember discussing this in 2018. 
You're definitely correct @cmadjar that it's not actually imaging, and that's why in 2018 we decided it gets its own menu.
My personal feeling is it should be under Imaging (👍 @johnsaigle ) because: 
(a) it's close enough to imaging, in terms of the context of the researchers, and compared to the testing/visit workflows used for behavioural/genomic...  
Plus, Loris Menu "real estate" in 2 senses:
(b) having a menu with just one item, as well as:
(c) having a super-wide menu item -- this genuinely messes up the Loris menu look at med-large screens and smaller

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param integer $physiologicalFileID The PhysiologicalFileID to be loaded
+     */
+    function __construct($physiologicalFileID)
```
```suggestion
    function __construct(int $physiologicalFileID)
```

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+     */
+    function __construct($physiologicalFileID)
+    {
+        $db     =& Database::singleton();
```
```suggestion
        $db     = \NDB_Factory::singleton()->database();
```
As per @AlexandraLivadas's [presentation](https://alexandralivadas.github.io/assets/GSOC_Presentation.pdf) (page 9) we should be using the Factory instead of the singleton directly. This allows for much easier creation of unit tests.

christinerogers replied:
Good callback, @johnsaigle 🎩

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+    function getParameter($parameterName)
+    {
+        return $this->fileData[$parameterName] ??
+            ($this->parameters[$parameterName] ?? '');
```
```suggestion
            $this->parameters[$parameterName] ?? 
            '';
```

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param string $parameterName The parameter to be retrieved
+     *
+     * @return mixed The value of the parameter
```
```suggestion
     * @return string The value of the parameter
```

This value comes from the DB so I think it's safe to use string here.

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return mixed The value of the parameter
+     */
+    function getParameter($parameterName)
```
```suggestion
    function getParameter(string $parameterName)
```

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+            WHERE ppf.ParameterTypeID=pt.ParameterTypeID AND
+            PhysiologicalFileID=:PFID";
+        $parameterRaw = $db->pselect($query, $params);
+        foreach ($parameterRaw AS $row) {
```
```suggestion
        foreach ($parameterRaw AS $name => $value) {
```

johnsaigle commented on modules/electrophysiology_browser/ajax/BIDSFile.class.inc:
```diff
@@ -0,0 +1,62 @@
+            PhysiologicalFileID=:PFID";
+        $parameterRaw = $db->pselect($query, $params);
+        foreach ($parameterRaw AS $row) {
+            $this->parameters[$row['Name']] = $row['Value'];
```
```suggestion
            $this->parameters[$name] = $value;
```

Not a necessary change but it's a little cleaner.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
Thank you! Will be added in the next commit.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+ * have the view site permission and belong to candidates session site
+ * are able to fetch the eeg session data.
+ */
+$user =& \User::singleton();
```
Same thing here, as above. Factory should be used instead of the User::singleton function.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+)
+{
+    header('HTTP/1.1 403 Forbidden');
+    exit;
```
I think we're trying to move away from the ajax exiting structure for modules. It involves a bit of refactoring so it probably shouldn't block this PR (especially since I can see you aren't the one who wrote it @cmadjar). But we should start not allowing this for new modules... at some point.

cmadjar replied:
True. Not sure how to refactor that for now though... Can be a separate PR maybe?

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+ */
+$user =& \User::singleton();
+$timePoint =& \TimePoint::singleton($_REQUEST['sessionID']);
+if (!$user->hasPermission('electrophysiology_browser_view_allsites')
```
This could probably be rewritten to be more clear. It should at least have a comment explaining the permissions.

cmadjar replied:
I added a comment explaining it. Hopefully that helps clarifying. Not sure how to clarify the code itself better...

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+
+echo json_encode($response);
+
+// TODO Query code here to echo JSON data to the user.
```
Is this done by line 39? If so, this should be deleted.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+//echo json_encode(array('hello'=>'world'));
+// TODO Query code here to echo JSON data to the user.
+//echo json_encode(array('hello'=>'world'));
+function getSessionData($sessionID)
```
```suggestion
function getSessionData( string $sessionID)
```

The functions in this file probably won't pass Travis... They need PHPDocs.

cmadjar replied:
hehe. Indeed, it did not pass Travis and I added the docs in the last commit ;).

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+{
+    $db = \Database::singleton();
+
+    $query = 'SELECT DISTINCT(pf.SessionID) FROM physiological_file pf LEFT
```
It would be nice to have this query split on different lines, similar to our row provisioner classes.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+    $response['database']    = array_values(getFilesData($sessionID));
+    $response['sessions']    = $sessions;
+    $currentIndex            = array_search($sessionID,$sessions);
+    $response['nextSession'] = isset($sessions[$currentIndex+1]) ?
```
```suggestion
    $response['nextSession'] = $sessions[$currentIndex+1] ?? '';
```

Same with prevSession below.

cmadjar replied:
Sweet! :)

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+    $response['sessions']    = $sessions;
+    $currentIndex            = array_search($sessionID,$sessions);
+    $response['nextSession'] = isset($sessions[$currentIndex+1]) ?
+        $sessions[$currentIndex+1] : '';
```
```suggestion
```

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+function getSubjectData($sessionID)
+{
+    $subjectData = array();
+    $timePoint   =& \TimePoint::singleton($sessionID);
```
```suggestion
    $timePoint   =& \TimePoint::singleton($sessionID);
```

Factory instead. I won't keeping on each place but it should be fixed in all cases.

cmadjar replied:
not sure I understand your comment. Could you explain it via Slack or the PR when you have some time? Thanks!

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,203 @@
+        );
+
+        $fileSummary['name'] = $fileName;
+        $fileSummary['task']['frequency']['sampling'] = $physiologicalFileObj->getParameter('SamplingFrequency');
```
This part makes me want to cry. 😿 PTSD from the statistics module.

Is there anyway to refactor this into a function?

johnsaigle commented on modules/electrophysiology_browser/php/eeg_visualization.class.inc:
```diff
@@ -0,0 +1,56 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

johnsaigle commented on modules/electrophysiology_browser/php/eeg_visualization.class.inc:
```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * Created by IntelliJ IDEA.
```
All this should be altered to match the normal way we declare PHP class and file docs. it won't pass Travis.

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_browser.class.inc:
```diff
@@ -0,0 +1,131 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
Done in the next commit. Should this be on top of each php script?

johnsaigle replied:
Yep!

johnsaigle commented on modules/electrophysiology_browser/php/eeg_visualization.class.inc:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @return bool whether the user hass access
+     */
+    function _hasAccess()
```
```suggestion
    function _hasAccess(): bool
```

johnsaigle commented on modules/electrophysiology_browser/php/eeg_visualization.class.inc:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Determine whether the user has permission to view this page
+     *
+     * @return bool whether the user hass access
```
```suggestion
     * @return bool whether the user has access
```

🐍

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_session.class.inc:
```diff
@@ -0,0 +1,74 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiologybrowserrow.class.inc:
```diff
@@ -0,0 +1,66 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiologybrowserrow.class.inc:
```diff
@@ -0,0 +1,66 @@
+     *                     returns
+     * @param integer $cid The centerID affiliated with this row.
+     */
+    public function __construct(array $row, $cid)
```
```suggestion
    public function __construct(array $row, int $cid)
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiologybrowserrow.class.inc:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return integer The CenterID
+     */
+    public function getCenterID()
```
```suggestion
    public function getCenterID(): int
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiologybrowserrowprovisioner.class.inc:
```diff
@@ -0,0 +1,80 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/module.class.inc:
```diff
@@ -0,0 +1,28 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/php/module.class.inc:
```diff
@@ -0,0 +1,28 @@
+ * This serves as a hint to LORIS that this module is a real module.
+ * It does nothing but implement the module class in the module's namespace.
+ *
+ * PHP Version 5
```
```suggestion
 * PHP Version 7
```

cmadjar replied:
ooops ;)

johnsaigle commented:
@cmadjar In Alex's presentation, she explained that

```php
$db = \Database::singleton();
```

results in difficult unit testing.

Instead, we should be using:

``php
$db = \NDB_Factory::singleton()->database();
```

This is because the Factory class does extra checking to figure out whether the code is running in a testing environment or not.

I believe the same goes for config, timepoint, ..., etc.

So in my review I was recommending that each of these singletons be declared using Factory so that it'll be easier to eventually write unit tests.

cmadjar commented:
@johnsaigle thank you for the clarification! The last commit replaces all occurence with calls to the NDB_Factory. Thanks!!

driusan commented on .eslintignore:
```diff
@@ -1,5 +1,6 @@
 node_modules/*
 modules/*/js/*.js
+modules/*/js/components/*.js
```
```suggestion
modules/*/js/*/*.js
```

I don't think there's any subdirectory of compiled js files that are supposed to be analyzed

cmadjar commented:
@driusan @kongtiaowang @johnsaigle I don't any error on the Travis log but it is still failing Travis. Could you enlighten me on what I am missing please? Thank you!

driusan commented:
It looks like it's freezing/timing out after building a docker image and trying to start mysql..

```
WARNING: Image for service integration-tests was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.

Waiting for mysqld...

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```

cmadjar commented:
@driusan Thank you! Is there anything that can be done to resolve that or just restart the build again? (Second time it is failing in a row)

driusan commented:
I can't think of anything that would have changed lately regarding that.. I restarted it just in case

driusan commented:
Looks like it's stuck again.. I don't understand Travis on this one either, I don't see your PR doing anything that should affect mysql starting.

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+                         pf.FilePath 
+                       FROM 
+                         physiological_file pf ';
+    //WHERE SessionID=:SID
```
Why is this commented?

cmadjar replied:
removed cause the WHERE statement is integrated in the if else below.

xlecours commented on modules/electrophysiology_browser/jsx/columnFormatter.js:
```diff
@@ -0,0 +1,52 @@
+ * @param {arrray} rowHeaders - array of table headers (column names)
+ * @return {*} a formated table cell for a given column
+ */
+function formatColumn(column, cell, rowData, rowHeaders) {
```
is this file used anywhere?

cmadjar replied:
apparently not. Removed the file.

kongtiaowang commented:
restart it to debug Travis

kongtiaowang commented on SQL/0000-00-01-Permission.sql:
```diff
@@ -102,6 +102,8 @@ INSERT INTO `permissions` VALUES
     (53,'publication_view', 'Publication - Access to module', 2),
     (54,'publication_propose', 'Publication - Propose a project', 2),
     (55,'publication_approve', 'Publication - Approve or reject proposed publication projects', 2);
```
It should be a Comma at the end of this line, not a Semicolon, which is the reason Travis failed.

cmadjar replied:
thank you @kongtiaowang! Should be fixed in the last commit.

driusan replied:
Shouldn't that result in a syntax error while sourcing the SQL, and not MySQL freezing?

cmadjar replied:
yeah, I agree, this is weird... But this is passing Travis now, YEAH!!!!

kongtiaowang commented on SQL/0000-00-02-Menus.sql:
```diff
@@ -35,9 +35,10 @@ INSERT INTO LorisMenu (Label, OrderNumber) VALUES
-     ('Reports', 4),
-     ('Tools', 5),
-     ('Admin', 6);
+     ('Electrophysiology', 4)
```
missing a comma here

cmadjar replied:
ooops... Should be fixed in the next commit.

johnsaigle commented on .eslintignore:
```diff
@@ -1,5 +1,6 @@
 node_modules/*
 modules/*/js/*.js
+modules/*/js/*/*.js
```
Maybe lines 2 and 3 could just be `modules/*/js/*`?

cmadjar replied:
removed the `.js` at the end of the line 2 and 3.

johnsaigle commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

johnsaigle commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

johnsaigle replied:
This should still be applied

cmadjar replied:
oooops... Sorry I missed it...

johnsaigle commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @return mixed The value of the parameter
+     */
+    function getParameter($parameterName)
```
```suggestion
    function getParameter(string $parameterName)
```

cmadjar replied:
ok. Done in the next commit.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+    $response['database']    = array_values(getFilesData($sessionID));
+    $response['sessions']    = $sessions;
+    $currentIndex            = array_search($sessionID, $sessions);
+    $response['nextSession'] = isset($sessions[$currentIndex+1]) ?? '';
```
```suggestion
    $response['nextSession'] = $sessions[$currentIndex+1] ?? '';
```

`??` is syntactic sugar for an if-isset-else check so this is redundant

cmadjar replied:
Thanks! Done in the last commit.

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+    $response['sessions']    = $sessions;
+    $currentIndex            = array_search($sessionID, $sessions);
+    $response['nextSession'] = isset($sessions[$currentIndex+1]) ?? '';
+    $response['prevSession'] = isset($sessions[$currentIndex-1]) ?? '';
```
```suggestion
    $response['prevSession'] = $sessions[$currentIndex-1] ?? '';
```

cmadjar replied:
same, done in the last commit

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+ */
+$user      = \NDB_Factory::singleton()->user();
+$timePoint = \NDB_Factory::singleton()->timepoint(
+    intval($_REQUEST['sessionID'])
```
Could you update this file to use sessionID consistently? It's being used as both an int and a string in this file.

Linking #4369 because it would help in a case like this.

cmadjar replied:
done in the next commit

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+ *
+ * @return array with the subject information
+ */
+function getSubjectData($sessionID)
```
```suggestion
function getSubjectData($sessionID): array
```

Also add int or string to sessionID, depending on previous comment.

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+ *
+ * @return array with the file collection
+ */
+function getFilesData($sessionID)
```
```suggestion
function getFilesData($sessionID): array
```

Also add int or string to sessionID, depending on previous comment.

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,327 @@
+ * @return array array with the path to the different files associated to the
+ *               electrophysiology file
+ */
+function getDownloadlinks($physiologicalFileID, $physiologicalFile)
```
```suggestion
function getDownloadlinks(int $physiologicalFileID, string $physiologicalFile): array
```

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_session.class.inc:
```diff
@@ -0,0 +1,76 @@
+                || ((in_array(
+                    $timePoint->getData('CenterID'),
+                    $user->getData('CenterIDs')
+                ))
```
We should create a Utility function for this idiom. It's used in a lot of places and doesn't look very good in the code.

cmadjar replied:
sounds like a great plan to create a Utility function for this. Should this be done in a separate though since it would affect multiple modules? Let me know your thoughts on this.

cmadjar commented on jsx/StaticDataTable.js:
```diff
@@ -51,6 +51,9 @@ class StaticDataTable extends Component {
       if (this.state.Hide.defaultColumn) {
         $('#dynamictable').find('tbody td:eq(0)').hide();
       }
+      if (this.state.Hide.defaultColumn) {
```
no idea. Removed duplicate.

cmadjar commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @param integer $physiologicalFileID The PhysiologicalFileID to be loaded
+     */
+    function __construct($physiologicalFileID)
```
ok. Done in the last commit

cmadjar commented on modules/electrophysiology_browser/jsx/electrophysiology_session_view.js:
```diff
@@ -0,0 +1,355 @@
+ */
+window.onload = function() {
+  const sidebarContent = (
+    <SidebarContent previous={'hello'} next={'okay'}/>
```
nice words? ;)

cmadjar replied:
replaced them by 'previous' and 'next'. Not really sure what they are used for. @maltheism ?

cmadjar commented on modules/electrophysiology_browser/php/electrophysiology_session.class.inc:
```diff
@@ -0,0 +1,76 @@
+ * @link       https://github.com/aces/Loris
+ */
+
+class Electrophysiology_Session extends \NDB_Form
```
indeed.

cmadjar commented on modules/electrophysiology_browser/php/electrophysiology_visualization.class.inc:
```diff
@@ -0,0 +1,80 @@
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+class Electrophysiology_Visualization extends \NDB_Form
```
so true... I had to remove the vis as it was not working for this PR. Forgot to remove this class at the same time...

johnsaigle commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,326 @@
+    $subjectData['dob']         = $candidate->getCandidateDoB();
+    $subjectData['sex']         = $candidate->getCandidateSex();
+    $subjectData['subproject']  = $timePoint->getData('SubprojectTitle');
+    $subjectData['output_type'] = $_REQUEST['outputType'];
```
Is this value being displayed on the front-end? If so, this value should be sanitized:

```suggestion
    $subjectData['output_type'] = htmlentities($_REQUEST['outputType'], ENT_QUOTES, "UTF-8");
```

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_browser.class.inc:
```diff
@@ -0,0 +1,131 @@
+     * the UserSiteMatch filter.
+     *
+     * @return bool always true
+     * @throws \DatabaseException
```
```suggestion

```

cmadjar replied:
ok. Done

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_browser.class.inc:
```diff
@@ -0,0 +1,131 @@
+            }
+        } else {
+            // allow only to view own site data
+            $siteIDs = $user->getData('CenterIDs');
```
```suggestion
            $siteIDs = $user->getCenterIDs();
```

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_browser.class.inc:
```diff
@@ -0,0 +1,131 @@
+            foreach ($list_of_sites as $key => $site) {
+                unset($list_of_sites[$key]);
+                $list_of_sites[$site] = $site;
+            }
```
```php
$siteNames = array_values(\Utility::getSiteList());
$list_of_sites = array_combine($siteNames, $siteNames);
```

cmadjar replied:
ok. Done

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiology_session.class.inc:
```diff
@@ -0,0 +1,73 @@
+        );
+
+        return ($user->hasPermission('electrophysiology_browser_view_allsites')
+                || ($user->hasCenter($timePoint->getData('CenterID'))
```
```suggestion
                || ($user->hasCenter($timePoint->getCenterID())
```

cmadjar replied:
done

cmadjar commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @param string $parameterName The parameter to be retrieved
+     *
+     * @return mixed The value of the parameter
```
ok. Changed

cmadjar commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,326 @@
+if (!$user->hasPermission('electrophysiology_browser_view_allsites')
+    && !((in_array($timePoint->getData('CenterID'), $user->getData('CenterIDs')))
+    && $user->hasPermission('electrophysiology_browser_view_site'))
+) {
```
done

maltheism commented:
Hi @cmadjar, I think I'm doing something wrong and maybe you can help. I'm trying to manual test your PR and I'm starting from a fresh install with rasinbread data. The issue I'm having is not seeing the new electrophysiology_browser module after running the usual mysql setup.

I used the below sql to setup everything with rasinbread.
```
cat SQL/0000-00-00-schema.sql \
    SQL/0000-00-01-Permission.sql \
    SQL/0000-00-02-Menus.sql \
    SQL/0000-00-03-ConfigTables.sql \
    SQL/0000-00-04-Help.sql \
    SQL/0000-00-05-ElectrophysiologyTables.sql \
    raisinbread/instruments/instrument_sql/aosi.sql \
    raisinbread/instruments/instrument_sql/bmi.sql \
    raisinbread/instruments/instrument_sql/medical_history.sql \
    raisinbread/instruments/instrument_sql/mri_parameter_form.sql \
    raisinbread/instruments/instrument_sql/radiology_review.sql \
    test/test_instrument/testtest.sql \
    raisinbread/RB_files/*.sql | mysql -u root -p -h localhost major
```

johnsaigle commented:
@maltheism It looks like @cmadjar hasn't added the new tables to Raisinbread yet so I added a label to indicate that.

If you want to test in the mean time, you can source the SQL patch manually that is included in this PR.

cmadjar commented:
@johnsaigle @maltheism Raisin bread updated!

maltheism commented:
Hi @cmadjar, I just manually tested it and I found one small questionable issue. When a user clicks on a button for downloading a file and the file doesn't exist. No LORIS Exception is thrown and a white page is just displayed.
Otherwise everything looks beautiful!

johnsaigle commented:
@maltheism Could you print the message from the error log that is generated when you get that blank page? It'll help isolate the problematic code

maltheism commented:
The error_log content:

> [Tue Oct 29 15:31:37.096876 2019] [php7:notice] [pid 94068] [client ::1:56099] ERROR: Requested file /data-raisinbread//bids_imports/Face13_BIDSVersion_1.1.0/sub-OTT166/ses-V1/eeg/sub-OTT166_ses-V1_task-faceO_eeg.tgz does not exist, referer: http://localhost/electrophysiology_browser/electrophysiology_session/?sessionID=166&outputType=raw&backURL=/electrophysiology_browser/

cmadjar commented:
Note to self: 
- fix travis
- fix download with the new FileDownload.php file. For some reason, it does not seem to be found when I click on the buttons. 

(I welcome any idea to fix the last point :)) @maltheism, @johnsaigle, @HenriRabalais, @driusan ? Thanks :)

johnsaigle commented on modules/electrophysiology_browser/ajax/FileDownload.php:
```diff
@@ -0,0 +1,48 @@
+<?php
```
```suggestion
<?php declare(strict_types=1);
```

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/ajax/FileDownload.php:
```diff
@@ -0,0 +1,48 @@
+ * This ensures that the file exists and the user is logged in to
+ * LORIS before trying to return the file to the user.
+ *
+ * PHP Version 5
```
```suggestion
 * PHP Version 7
```

cmadjar replied:
done

johnsaigle commented on modules/electrophysiology_browser/ajax/FileDownload.php:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @category Loris
+ * @package  Media
+ * @author   Alex I. <ailea.mcin@gmail.com>
```
welcome back, Alex!

cmadjar replied:
haha. Oooops. Fixed.

johnsaigle commented on modules/electrophysiology_browser/ajax/FileDownload.php:
```diff
@@ -0,0 +1,48 @@
+if (!file_exists($filePath)) {
+    error_log("ERROR: File $filePath does not exist");
+    header("HTTP/1.1 404 Not Found");
+    exit(5);
```
```suggestion
    exit;
```

cmadjar replied:
done

cmadjar commented:
@johnsaigle Security issue should be fixed in the last commit.

johnsaigle commented:
@cmadjar Is this still Blocked? Should it go into 22 or no?

cmadjar commented:
@johnsaigle I removed the blocked label and rebased it to major. It could go to 22.0 or go to the next release. No strong preference on this one. @driusan @ridz1208?

cmadjar commented:
Note to self: for some reason the viewSession page is not loading anymore. React problem. Will need to look into it on Monday

johnsaigle commented on modules/electrophysiology_browser/php/electrophysiologybrowserrow.class.inc:
```diff
@@ -37,10 +38,11 @@ class ElectrophysiologyBrowserRow implements \LORIS\Data\DataInstance
      *                     returns
      * @param integer $cid The centerID affiliated with this row.
```
I think you'll need to add a PHPDoc here for the new param or else Travis will complain.

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ */
+
+require_once 'ElectrophysioFile.class.inc';
+
```
There should be a check to see if `$_REQUEST['sessionID']` is set and is a number and return a `400 Bad Request` if it's not the case.

cmadjar replied:
ok. Done

johnsaigle commented:
Looks like Travis is having a hard time getting started.. not sure why

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ *
+ * @return array with the subject information
+ */
+function getSubjectData(int $sessionID)
```
$sessionID is not required. The function could use `$timePoint` defined at line 26.

cmadjar replied:
ok

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+    $subjectData['dob']         = $candidate->getCandidateDoB();
+    $subjectData['sex']         = $candidate->getCandidateSex();
+    $subjectData['subproject']  = $timePoint->getData('SubprojectTitle');
+    $subjectData['output_type'] = htmlentities(
```
This is weird. Why would the server need to return a value provided by the user?

cmadjar replied:
weird indeed... Removed it.

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ *
+ * @return array with the file collection
+ */
+function getFilesData(int $sessionID)
```
Again, sessionID not required. There is a timePoint already instantiated at line 26.

cmadjar replied:
it is required to have for the query. Otherwise I could pass $timePoint but then I would have to add a line
```
$sessionID = $timepoint->getSessionID;
```
in this case, I am not sure of the benefit of this. Maybe there is something I don't get. Very likely, lol

xlecours replied:
right

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ */
+
+require_once 'ElectrophysioFile.class.inc';
+
```
$_REQUEST['outputType'] shoudl also be checked early. Additionally, make sure the value is raw, derivatives or all_types. Otherwise, getFilesData will return all types even where an invalid type is provided.

cmadjar replied:
ok

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+        $query .= 'LEFT JOIN physiological_output_type pot ON ';
+        $query .= 'pf.PhysiologicalOutputTypeID=pot.PhysiologicalOutputTypeID ';
+        $query .= 'WHERE SessionID=:SID ';
+        if ($outputType == 'raw') {
```
if the output_type is validated at the start of the file, this can be streamline to `$query .= 'AND pot.OutputTypeName = "' . $outputType . '"';`

cmadjar replied:
done!

cmadjar commented on modules/electrophysiology_browser/ajax/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,74 @@
+            $this->fileData[$key] = $value;
+        }
+
+        $query        = "SELECT 
```
ok. Done

cmadjar commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ * @link     https://github.com/aces/Loris-Trunk
+ */
+
+require_once 'ElectrophysioFile.class.inc';
```
@xlecours should I add a blocked label and wait for the EEG API PR to be merged? Maybe there is some code there that I could reuse here? What do you think?

cmadjar commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,336 @@
+ * are able to fetch the electrophysiology session data.
+ */
+$user      = \NDB_Factory::singleton()->user();
+$timePoint = \NDB_Factory::singleton()->timepoint(
```
ok

cmadjar commented:
@xlecours I believe this is ready for another review. Thank you!

PapillonMcGill commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+## Intended Users
+
+The primary types of users are:
+1. Electrophysiology researchers who want to know details about the inserted datasets
```
```suggestion
1. Electrophysiology researchers who want to know details about the inserted datasets.
```

PapillonMcGill commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+## Download
+
+You can download all the files related to a recording (channel information,
+electrode information, task event information, the actual recording...)
```
```suggestion
electrode information, task event information, the actual recording...).
```

PapillonMcGill commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+sufficient to provide access to view data in the module.
+
+electrophysiology_browser_view_allsites
+  - This permission gives the user access to all electrophysiology datasets present in the database
```
```suggestion
  - This permission gives the user access to all electrophysiology datasets present in the database.
```

PapillonMcGill commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+  - This permission gives the user access to all electrophysiology datasets present in the database
+  
+electrophysiology_browser_view_site
+  - This permission gives the user access to electrophysiology datasets from their own site(s) only
```
```suggestion
  - This permission gives the user access to electrophysiology datasets from their own site(s) only.
```

xlecours commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,365 @@
+    return false;
+}
+
+$timePoint = \NDB_Factory::singleton()->timepoint(
```
This line can be removed. `$timePoint` is initialized at line #52

cmadjar commented:
@PapillonMcGill I committed your suggestions. Thanks!

johnsaigle commented:
@kongtiaowang Can you update your review?

PapillonMcGill commented:
@cmadjar One point missing but approved anyway as not a major thing. Would be great if you can update.

cmadjar commented:
@PapillonMcGill @xlecours your last comments were integrated in the last commit. Thank you for your reviews!

driusan commented on .eslintignore:
```diff
@@ -1,5 +1,6 @@
 node_modules/*
-modules/*/js/*.js
+modules/*/js/*
+modules/*/js/*/*
```
Can you add a comment explaining that eslint is being ignored on the compiled javascript because it's run on the uncompiled jsx directory? At least, I hope that's why this is here..

cmadjar replied:
done

driusan commented on SQL/New_patches/2018-05-18_adding_physiological_all_sites_permissions.sql:
```diff
@@ -0,0 +1,39 @@
+    ) VALUES (
+      (SELECT ID FROM users WHERE UserID='admin'),
+      (SELECT permID FROM permissions WHERE code='electrophysiology_browser_view_allsites')
+    );
```
Not going to block over it because the performance impact is so minimal for a one-off patch, but I think it would be better  to write this as an `INSERT INTO (UserID, PermID) SELECT xxx` statement with a join.

driusan commented on SQL/New_patches/2020-01-20_adding_electrophysiology_browser_to_modules_table.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO modules (Name, Active) VALUES ('electrophysiology_browser', 'Y');
+
```
Also not a big deal because the PR was sent before the script was merged, but we should probably just direct people to use the tools/manage_modules script in the release notes.

driusan commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+NOT in scope at the moment:
+
+Visualization of channel signal data and electrode positions for each electrophysiology dataset. 
+(These are in development and will be added in future.)
```
The last sentence implies that it is, in fact, within the scope of what the module should cover..

driusan replied:
This doesn't seem to be resolved

cmadjar replied:
I am not sure to understand that one. The visualization is in the scope of the module but no one is there to develop it so it is not there yet. Should I just remove the last two sentences and we'll add them back when the vis will be developed?

driusan replied:
Oh, sorry, I misread based on a previous iteration of the document..

But 'Visualization of channel signal data and electrode positions for each electrophysiology dataset.' also doesn't make sense as a sentence, so you can either update the sentence to fix the grammar or remove it..

driusan commented on modules/electrophysiology_browser/ajax/get_electrophysiology_session_data.php:
```diff
@@ -0,0 +1,362 @@
+<?php declare(strict_types=1);
+/**
+ * Get Electrophysiology Session Data.
```
new modules shouldn't be using the `ajax` directory at all, this should be converted to a real URL endpoint now that LORIS has a way to route requests properly. It should be something like `$lorisBase/electrophysiology_browser/eeg_session/$sessionid` (for some reasonable unique definition of `$sessionid`)

driusan commented on modules/electrophysiology_browser/php/electrophysiology_browser.class.inc:
```diff
@@ -0,0 +1,148 @@
+{
+
+    public $AjaxModule   = true; // stops from running the query twice
+    public $skipTemplate = true; // stops from looking for a smarty template
```
both of these are done by the DataFrameworkMenu constructor, these lines should just be deleted

driusan commented on modules/electrophysiology_browser/php/models/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,85 @@
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class ElectrophysioFile implements \LORIS\Data\DataInstance
```
I don't get why this PR makes a distinction between a "ElectrophysiologyBrowserRowProvisioner" DataInstance and a "ElectrophysioFile" DataInstance (which serializes to JSON as the empty string?). There seems to be one row per file in the table. Why not implement toJSON on the ElectrophysioFile and use that as the type of DataInstance returned by the provisioner?

It seems to be taking the hack that was used to add models to existing data tables and using that as the "right" way of doing things rather than writing it from scratch to re-use the models for both the dataframework and the code.

cmadjar replied:
GOUP BY

cmadjar commented on modules/electrophysiology_browser/README.md:
```diff
@@ -0,0 +1,40 @@
+inserted into LORIS from a BIDS-format collection. Derived or processed electrophysiology
+datasets can also be accessed via this module.
+
+NOT in scope at the moment:
```
```suggestion

```

cmadjar commented:
@driusan I'll fetch on Monday. Too brain dead to finish that last part. Will let you know when ready for another review. Thanks again for your help today! :)

cmadjar commented:
@driusan replaced the ajax with fetch API.

cmadjar commented:
@kongtiaowang I have no idea why Travis is failing. I don't see any error message but that could just be me not able to spot the error ;).

kongtiaowang commented on SQL/0000-00-01-Permission.sql:
```diff
@@ -103,6 +103,8 @@ INSERT INTO `permissions` VALUES
     (55,'publication_propose', 'Publication - Propose a project', 2),
     (56,'publication_approve', 'Publication - Approve or reject proposed publication projects', 2),
     (57, 'candidate_dob_edit', 'Edit dates of birth', 2);
```
line 105, change semicolon to comma. " ; " => " , "
It should pass travis after changing.

cmadjar replied:
@kongtiaowang thank you!! Fixed.

cmadjar commented:
Wouhou, passing Travis, @driusan I believe the ball is in your court now :)

driusan commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -0,0 +1,386 @@
+      'download_fdt_file',
+    ];
+    for (let i=0; i<this.state.data.downloads.length; i++) {
+      console.log(this.state.data.downloads[i]);
```
Leftover console.log?

cmadjar replied:
oooops. Will be removed in the next commit

driusan commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -0,0 +1,386 @@
+   *
+   */
+  componentDidUpdate(prevProps, prevState) {
+  }
```
Why is this empty function here?

cmadjar replied:
oooops. Will be removed in the next commit

driusan commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -0,0 +1,386 @@
+          <div className={'row'}>
+            {/* <div className={'col-sm-4'} style={stylesFile.div.container.details}> */}
+              {/* ..insert head image here.. */}
+            {/* </div> */}
```
There shouldn't be commented out code

cmadjar replied:
agreed :)

driusan commented on modules/electrophysiology_browser/jsx/electrophysiologyBrowserIndex.js:
```diff
@@ -0,0 +1,162 @@
+  constructor(props) {
+    super(props);
+
+    // loris.hiddenHeaders = ['Session ID'];
```
More commented out code

cmadjar replied:
done

driusan commented on modules/electrophysiology_browser/jsx/electrophysiologyBrowserIndex.js:
```diff
@@ -0,0 +1,162 @@
+    // Bind component instance to custom methods
+    this.fetchData = this.fetchData.bind(this);
+    // this.updateFilter = this.updateFilter.bind(this);
+    // this.resetFilters = this.resetFilters.bind(this);
```
And here

cmadjar replied:
done

driusan commented on modules/electrophysiology_browser/jsx/electrophysiologySessionView.js:
```diff
@@ -0,0 +1,356 @@
+   * @param {object} prevState
+   */
+  componentDidUpdate(prevProps, prevState) {
+    console.log('componentDidUpdate');
```
I don't think this should be here, and then removing it makes another empty function that should be removed..

cmadjar replied:
removed

driusan commented on modules/electrophysiology_browser/php/models/ElectrophysioFile.class.inc:
```diff
@@ -0,0 +1,86 @@
+
+        $db = \NDB_Factory::singleton()->database();
+
+        $query    = "SELECT * 
```
This should be an explicit list of files, not a `SELECT *`

cmadjar replied:
ok. Done

cmadjar commented:
@driusan Your comments have been taken care of. Thanks!

cmadjar commented:
@driusan Edit to the README done :). Back to you.

PS: subliminal message: merge, merge, merge ;)

driusan commented:
Closing and re-opening this to try and get it to start Travis..

cmadjar commented:
![image](https://user-images.githubusercontent.com/1402456/74254537-92e1f300-4cbe-11ea-9563-82a355d319fb.png)
